### PR TITLE
Add US-SEC parser for Seminole Electric Cooperative in Florida

### PR DESCRIPTION
**Pull request by snarfed**
combines gas + coal data from EIA with solar data from apps.seminole.coop. still needs geojson. lots of background and details in #1713. example data:

```py
[{'datetime': datetime.datetime(2019, 7, 14, 23, 0, tzinfo=<UTC>),
  'production': {'solar': 0.6502, 'unknown': None},
  'source': 'eia.org, apps.seminole.coop',
  'storage': {},
  'zoneKey': 'US-SEC'},
  ...
]
```

both data sources usually lag by a day or two, sometimes much more. for example, right now EIA's data stops on 2019-07-09, almost 10d ago. :slightly_frowning_face:

minor quirk: this parser reuses code from both the EIA and ENTSOE parsers, which there isn't much other precedent for so far. i've added a TODO to move `ENTSOE.merge_production_outputs` to `lib.utils`, but the EIA reuse might be ok as is? see what you think.

thanks in advance for reviewing, and for all your support so far!

### CODE REVIEW

File: config/zones.json [hunk 1]
```diff
@@ -3708,6 +3708,31 @@
     },
     "timezone": null
   },
+  "US-SEC": {
+    "_comment": "https://www.seminole-electric.com/facilities/generation/ lists two 650MW coal-fired plants, one 810MW gas plant, and a small 2.2MW solar farm. Hwoever, EIA combines coal and gas generation data, so we report those together as unknown",
```
Comment (corradio) [inline config/zones.json]: can you put this comment in the parser itself? This is the best place to put this IMO

File: parsers/EIA.py [hunk 2]
```diff
@@ -32,16 +36,28 @@
...
+
+def fetch_production(zone_key, session=None, target_datetime=None,
+                     logger=None):
+    return _fetch_production_or_consumption(
+        zone_key, PRODUCTION[zone_key], session=session,
+        target_datetime=target_datetime, logger=logger)
+
 
-    series_id = DAY_AHEAD[zone_key]
+def _fetch_production_or_consumption(zone_key, series_id, session=None,
```
Comment (corradio) [inline parsers/EIA.py]: how do we know if it's a production or consumption?

File: parsers/ENTSOE.py [hunk 3]
```diff
@@ -857,6 +857,8 @@ def fetch_production(zone_key, session=None, target_datetime=None,
 }
 
 
+# TODO: generalize and move to lib.utils so other parsers can reuse it. (it's
```
Comment (corradio) [inline parsers/ENTSOE.py]: 👍

File: config/zones.json [hunk 4]
```diff
@@ -3708,6 +3708,31 @@
     },
     "timezone": null
   },
+  "US-SEC": {
```
Comment (corradio) [inline config/zones.json]: do we already have a geometry for this?

Comment (snarfed) [reply config/zones.json]: sadly, no. that's my next project!

File: config/zones.json (hunk 1 above)
Comment (snarfed) [reply config/zones.json]: good call! done.

File: parsers/EIA.py (hunk 2 above)
Comment (snarfed) [reply parsers/EIA.py]: good question. the series id, specifically `DF` for consumption forecast vs `NG` for production. i've added a docstring.

File: config/zones.json (hunk 4 above)
Comment (corradio) [reply config/zones.json]: ok, then let's wait to merge this PR for when the geometry is ready

File: config/zones.json (hunk 1 above)
Comment (corradio) [reply config/zones.json]: great! Then I think you can remove it from here.

File: config/zones.json (hunk 4 above)
Comment (snarfed) [reply config/zones.json]: sure. i may need a bit of help, though. details in https://github.com/tmrowco/electricitymap-contrib/issues/1713#issuecomment-513506553 . if you know who i should talk to, please point me their way!

**Comment by snarfed** [comment]
ok, this is close to ready for re-review! i've added geometry based on roughly half of Florida's counties. screenshot below; discussion in https://github.com/tmrowco/electricitymap-contrib/issues/1713#issuecomment-517704023 and onward.

two possible TODOs before merging this ([details](https://github.com/tmrowco/electricitymap-contrib/issues/1713#issuecomment-518012542)):
* merge county boundary lines so they're not visible
* fix bug in `topogen.sh` so that it correctly persists and caches `US-FL.geojson` instead of downloading again every time

image

**Comment by systemcatch** [comment]
Hey @snarfed you don't need to do a check for the file. Just put it in https://github.com/tmrowco/electricitymap-contrib/tree/master/web/third_party_maps that way it will get copied with all the other shapes on https://github.com/tmrowco/electricitymap-contrib/blob/master/web/topogen.sh#L28

**Comment by snarfed** [comment]
@systemcatch sure! agreed, that'd be easier. i didn't consider that originally because it's 3MB though. is that still ok to go into `third_party_maps`?

alternatively i could just check in the final 4-5KB of polygons, after compressing and quantizing to the `world.json` grid, but that would lose the provenance and steps to regenerate from the original Florida counties data.

**Comment by snarfed** [comment]
oh, i see that the two NACIS zip files in `web/build/` are bigger, and they're checked in. ok then! will do.

File: web/generate-geometries.js [hunk 5]
```diff
@@ -593,6 +606,7 @@ const zoneDefinitions = [
   { zoneName: 'US-PJM', type: 'states', countryId: 'USA', states: [
     'US.PA', 'US.NJ', 'US.MD', 'US.DE', 'US.VA', 'US.WV', 'US.OH', 'US.KY']},
   // { zoneName: 'US-RI', countryId: 'USA', stateId: 'US.RI', type: 'state' },
+  { zoneName: 'US-SEC', type: 'county', counties: ['ALACHUA', 'BAKER', 'BRADFORD', 'CITRUS', 'CLAY', 'COLUMBIA', 'DESOTO', 'DIXIE', 'GADSDEN', 'GILCHRIST', 'GLADES', 'HAMILTON', 'HARDEE', 'HENDRY', 'HERNANDO', 'HIGHLANDS', 'JEFFERSON', 'LAFAYETTE', 'LAKE', 'LEON', 'LEVY', 'LIBERTY', 'MADISON', 'MANATEE', 'OKEECHOBEE', 'OSCEOLA', 'PASCO', 'PUTNAM', 'SARASOTA', 'SUMTER', 'SUWANNEE', 'TAYLOR', 'UNION', 'WAKULLA']},
```
Comment (corradio) [inline web/generate-geometries.js]: can you put this line after `US-SC` so it is ordered alphabetically?

File: web/generate-geometries.js [hunk 6]
```diff
@@ -750,9 +767,14 @@ zoneFeatures = toListOfFeatures(zoneFeaturesInline);
 // Write unsimplified list of geojson, without state merges
 fs.writeFileSync('public/dist/zonegeometries.json', zoneFeatures.map(JSON.stringify).join('\n'));
 
-// Simplify all countries
+// Convert to TopoJSON
 const topojson = require('topojson');
 let topo = topojson.topology(zones);
+
+// merge Florida counties in US-SEC
```
Comment (corradio) [inline web/generate-geometries.js]: maybe add a more explicit description that points to the fact that we do this to avoid having inner lines?

Comment (snarfed) [reply web/generate-geometries.js]: agreed, good call, done.

File: web/generate-geometries.js (hunk 5 above)
Comment (snarfed) [reply web/generate-geometries.js]: oops, yes, done.

**Comment by corradio** [comment]
Let's see if it works!

**Pull request merged**